samza-core/src/main/java/org/apache/samza/coordinator/InputStreamRegexMonitor.java

```diff
@@ -0,0 +1,242 @@
+   */
+  public void stop() {
+    synchronized (lock) {
+      // We could also wait for full termination of the scheduler service, but it is overkill for
+      // our use case.
+      schedulerService.shutdownNow();
+
+      state = State.STOPPED;
+    }
+  }
+
+  private void monitorInputRegexes() {
+    log.debug("Running monitorInputRegexes");
+
+    try {
+      // obtain the list of SysStreams that match given patterns for all systems
+      Set<SystemStream> inputStreamsMatchingPattern = new HashSet<>();
+
+      // For each input system, for which we have a regex to monitor
+      for (String systemName : this.systemRegexesToMonitor.keySet()) {
+
+        try {
+          // obtain the list of SysStreams that match the regex for this system
+          // using the systemAdmin in the metadataCache
+          inputStreamsMatchingPattern.addAll(
+              JavaConverters.setAsJavaSetConverter(this.metadataCache.getAllSystemStreams(systemName))
+                  .asJava()
+                  .stream()
+                  .filter(x -> x.getStream().matches(this.systemRegexesToMonitor.get(systemName).pattern()))
+                  .collect(Collectors.toSet()));
+
+        } catch (UnsupportedOperationException e) {
+          log.error("UnsupportedOperationException while monitoring input regexes for system {}", systemName, e);
+        }
+      }
+
+      // if there is a stream that is in the input-Set but not in the streamsToMonitor
+      // since streamsToMonitor = task.inputs
+      if (!streamsToMonitor.containsAll(inputStreamsMatchingPattern)) {
+        log.info("New input system-streams discovered. InputStreamsMatchingPattern: {} but streamsToMonitor: {} ",
+            inputStreamsMatchingPattern, streamsToMonitor);
+
+        // invoke notify callback with new input streams
+        this.callbackMethod.onNewInputStreamsDiscovered(Sets.difference(inputStreamsMatchingPattern, streamsToMonitor));
+      } else {
+        log.info("No new input system-Streams discovered streamsToMonitor {} inputStreamsMatchingPattern {}",
+            streamsToMonitor, inputStreamsMatchingPattern);
+      }
+    } catch (Exception e) {
+      log.error("Exception while monitoring input regexes.", e);
+    }
+  }
+
+  /**
+   * For testing. Returns the metrics.
+   */
+  @VisibleForTesting
+  Map<String, Gauge<Integer>> getGauges() {
+    return gauges;
+  }
+
+  @VisibleForTesting
+  boolean isRunning() {
+    return state == State.RUNNING;
+  }
+
+  /**
+   * Wait until this service has shutdown. Returns true if shutdown occurred within the timeout
+   * and false otherwise.
+   * <p>
+   * This is currently exposed at the package private level for tests only.
+   */
+  @VisibleForTesting
+  boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return schedulerService.awaitTermination(timeout, unit);
+  }
+
+  private static class ThreadFactoryImpl implements ThreadFactory {
```

Comment:
Cleaner to use https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/ThreadFactoryBuilder.html instead of rolling our own implementation

Reply:
done